File: THREE/Math/Cylindrical.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using _Math = System.Math;

namespace THREE
{
    /// <summary>
    /// @author Mugen87 / https://github.com/Mugen87
    /// @author tengge / https://github.com/tengge1
    /// Ref: https://en.wikipedia.org/wiki/Cylindrical_coordinate_system
    /// </summary>
    public class Cylindrical
    {
        public double radius;
        public double theta;
        public double y;

        public Cylindrical(double radius = 1.0, double theta = 0, double y = 0)
        {
            this.radius = radius; // distance from the origin to a point in the x-z plane
            this.theta = theta; // counterclockwise angle in the x-z plane measured in radians from the positive z-axis
            this.y = y; // height above the x-z plane
        }

        public Cylindrical Set(double radius, double theta, double y)
        {
            this.radius = radius;
            this.theta = theta;
            this.y = y;

            return this;
        }

        public Cylindrical Clone()
        {
            return new Cylindrical().Copy(this);
        }

        public Cylindrical Copy(Cylindrical other)
        {
            this.radius = other.radius;
            this.theta = other.theta;
            this.y = other.y;

            return this;
        }

        public Cylindrical SetFromVector3(Vector3 v)
        {
            return this.setFromCartesianCoords(v.x, v.y, v.z);
        }

        Cylindrical SetFromCartesianCoords(double x, double y, double z)
        {
            this.radius = _Math.Sqrt(x * x + z * z);
            this.theta = _Math.Atan2(x, z);
            this.y = y;

            return this;
        }
    }
}

File: ShadowEditor.Server.Go/three/cylindrical.go
// Copyright 2017-2020 The ShadowEditor Authors. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.
//
// For more information, please visit: https://github.com/tengge1/ShadowEditor
// You can also visit: https://gitee.com/tengge1/ShadowEditor
//
// This package is translated from three.js, visit `https://github.com/mrdoob/three.js`
// for more information.

package three

import (
	"math"
)

/**
 * @author Mugen87 / https://github.com/Mugen87
 *
 * Ref: https://en.wikipedia.org/wiki/Cylindrical_coordinate_system
 *
 */

// NewCylindrical create a new Cylindrical
func NewCylindrical(radius, theta, y float64) *Cylindrical {
	return &Cylindrical{radius, theta, y}
}

type Cylindrical struct {
	Radius float64 // distance from the origin to a point in the x-z plane
	Theta  float64 // counterclockwise angle in the x-z plane measured in radians from the positive z-axis
	Y      float64 // height above the x-z plane
}

func (c Cylindrical) Set(radius, theta, y float64) *Cylindrical {
	c.Radius = radius
	c.Theta = theta
	c.Y = y
	return &c
}

func (c Cylindrical) Clone() *Cylindrical {
	return &Cylindrical{c.Radius, c.Theta, c.Y}
}

func (c Cylindrical) Copy(other *Cylindrical) *Cylindrical {
	c.Radius = other.Radius
	c.Theta = other.Theta
	c.Y = other.Y
	return &c
}

func (c Cylindrical) setFromVector3(v Vector3) *Cylindrical {
	return c.SetFromCartesianCoords(v.X, v.Y, v.Z)
}

func (c Cylindrical) SetFromCartesianCoords(x, y, z float64) *Cylindrical {
	c.Radius = math.Sqrt(x*x + z*z)
	c.Theta = math.Atan2(x, z)
	c.Y = y
	return &c
}
